pysb/bng.py

```diff
@@ -924,6 +924,47 @@ def _is_bool_expr(e):
     return isinstance(e, Boolean) and not isinstance(e, sympy.AtomicExpr)
 
 
+# BNG's expression syntax supports multiplication with a boolean, e.g.
+# `2 * (x > 0)`, but sympy does not. Our overall strategy is to convert
+# multiplication with a boolean expression into a BNG-equivalent Piecewise,
+# leaving other multiplications alone. First, we use a transformation function
+# during parsing to temporarily replace any '*' operators with a fake 'bngl_mul'
+# function call. Then, after parsing and evaluation, we perform a subexpression
+# replacement on the 'bngl_mul' calls that converts Boolean args to a Piecewise
+# and emits a normal Mul. This can't be done entirely in a parser transformation
+# because sympy hasn't yet parsed the expression to determine the AST structure
+# that lets us identify boolean subexpressions, and it can't be done entirely
+# post-parsing because multiplication of booleans raises an immediate error
+# during the evaluation step in parse_expr. This function is the parser
+# transformation.
+def _convert_mul(tokens, local_dict, global_dict):
+    res1 = sympy_parser._group_parentheses(_convert_mul)(tokens, local_dict, global_dict)
+    res2 = sympy_parser._apply_functions(res1, local_dict, global_dict)
+    res3 = _transform_mul(res2, local_dict, global_dict)
+    result = sympy_parser._flatten(res3)
+    return result
+
+
+# Helper function for _convert_mul.
+def _transform_mul(tokens, local_dict, global_dict):
+    # Strongly based on sympy.parsing.sympy_parser._transform_equals_sign.
+    result = []
+    if (tokenize.OP, '*') in tokens:
+        local_dict['__bngl_mul'] = sympy.Function('bngl_mul')
+        result.append((tokenize.NAME, '__bngl_mul'))
```

Comment:
Won't this break operator precedence as everything following will be executed after `bngl_mul`? I think it might be easier to directly transform `<` and `>` (and '=<' and '=>' if they are allowed in BNG) here as operator precedence wouldn't be an issue, it would't require any post-parsing processing and it would give compatibility other operands.

Reply:
I assume you mean converting all relational expressions to the corresponding 0 or 1 using a `Piecewise`? I don't think this works because we need to support BNG `if` expressions which we also convert to a sympy `Piecewise`, and I believe the conditions in a Piecewise needs to be booleans. I'll try it though.

Reply:
You're absolutely right on the operator precedence problem with the current implementation though. Not sure what I was thinking there. I'll add a test case to catch this.

Reply:
> I believe the conditions in a Piecewise needs to be booleans

I think I solved this by making the `bngl_if` implementation explicitly treat both `False` and `0` as false (and anything else as true):
```python
expr = expr.replace(
    sympy.Function('bngl_if'),
    lambda cond, t, f: sympy.Piecewise(
        (f, sympy.Eq(cond, False) | sympy.Eq(cond, 0)),
        (t, True),
    )
)
```